### PR TITLE
Tx execution with off-chain multiple signatures

### PR DESCRIPTION
**Pull request by graikos**
Implements an execution method in the smart contract that assumes multiple off-chain signatures for approval. The batch verification uses recursive proofs. Added some tests for both, but split the proof tests to run with `bun run test:proofs`, because they take much longer than the rest.

Important assumptions:
- The recursive proof exposes the hash of the last public key to sign. I suppose this is OK, since this public key will call the `execute` method, and their pub key will be visible in this TX.
- The recursive proof outputs the accumulated approval count, but DOES NOT perform a check. Also, it does not include any markers, just approvals. The calling contract must check the threshold etc.
- To prevent duplicate signatures, the recursive proofs must be created in ascending order (by pub key hash). Note that this does not affect the order at the time of signing, but the proof creation.
- The proposal must be already submitted on chain, with no prior approvals (strict). The contract assumes the signatures are enough to reach the threshold by themselves, so no implicit approval by the method caller, even if they're an owner.
- Takes into account the auto-approval on proposal, so expects `PROPOSED_MARKER + 1`

### CODE REVIEW

**Comment by mellowcroc** [comment]
> The recursive proof exposes the hash of the last public key to sign. I suppose this is OK, since this public key will call the execute method, and their pub key will be visible in this TX.

actually, I wonder if we need to go further and keep a running hash of the approver pub keys (e.g. H(H(approver1), approver2)...) just to keep some auditable trail. (txs in Safe)

> To prevent duplicate signatures, the recursive proofs must be created in ascending order (by pub key hash). Note that this does not affect the order at the time of signing, but the proof creation.

nice!

> The proposal must be already submitted on chain, with no prior approvals (strict).

as mentioned out-of-band, I think we should allow `executeTransferBatchSig` to run without a proposal being submitted beforehand. Basically, we just check whether the approval value is `EXECUTED_MARKER` or not and if not, run the verification flow. This will override any non-`executeTransferBatchSig` flows (e.g. approval value = PROPOSED_MARKER, PROPOSED_MARKER + 1, ...), but I think that should be ok.

File: contracts/src/BatchVerifyProgram.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        const approverHash = verifyApproval(input, sig, approver, ownerWitness);
+
+        // To prevent duplicate signatures, enforce ascending order of hashes of public keys
+        prev.publicOutput.approverHash.assertLessThan(approverHash);
+
+        return {
+          publicOutput: new BatchVerifyOutput({
+            approvalCount: prev.publicOutput.approvalCount.add(Field(1)),
+            approverHash: approverHash,
+            approverChain: Poseidon.hash([prev.publicOutput.approverChain, ...approver.toFields()])
```
Comment (teddav) [inline contracts/src/BatchVerifyProgram.ts]: should be `approverHash` instead of `...approver.toFields()`

Comment (graikos) [reply contracts/src/BatchVerifyProgram.ts]: I was going for a `H(H(H(p1), p2), p3)` chain, but I guess using `approverHash` would be the same and we already have it, so might be more efficient

File: contracts/src/BatchVerifyProgram.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        sig: Signature, approver: PublicKey, ownerWitness: MerkleMapWitness) {
+
+        prev.verify();
+
+        prev.publicInput.assertEquals(input);
+
+        const approverHash = verifyApproval(input, sig, approver, ownerWitness);
+
+        // To prevent duplicate signatures, enforce ascending order of hashes of public keys
+        prev.publicOutput.approverHash.assertLessThan(approverHash);
```
[the remark this answers is not in the document]
Comment (graikos) [reply contracts/src/BatchVerifyProgram.ts]: Signers don't have to sign in order, the final proof must be created in order over the signatures. If we assume the final user computes the entire proof, then this is not an issue. If we want each user that signs to also advance the proof by one step, then they would need to sign+prove in order.  The drawback is that we would need to have the list of approvers, and to verify the chain we would still need to have some kind of ordering (but could be easier, like sort by timestamp in front end and send in that order). Also, the front end would need to be in sync so that owners sign+prove sequentially. For reference, Safe uses the ascending order thing I have, but they don't have recursive proofs, so our use case is a bit different.

File: contracts/src/MinaGuard.ts [hunk 3]
```diff
@@ -251,25 +247,23 @@ export class MinaGuard extends SmartContract {
...
-    ownerWitness: MerkleMapWitness,
+    ownerWitness: OwnerWitness,
     proposer: PublicKey,
     signature: Signature,
     voteNullifierWitness: MerkleMapWitness,
     approvalWitness: MerkleMapWitness
   ) {
     // --- propose logic ---
-    const ownersRoot = this.getInitializedOwnersRoot();
+    const ownersRoot = this.getInitializedOwnersCommitment();
```
Comment (mellowcroc) [inline contracts/src/MinaGuard.ts]: ownersCommitment?

File: contracts/src/tests/governance.test.ts
Comment (mellowcroc) [inline contracts/src/tests/governance.test.ts]: could you add a test for checking that adding an owner above the maximum fails?

**Review by mellowcroc — COMMENTED** [review]
This looks great! Left a few comments/questions.

File: contracts/src/MinaGuard.ts [hunk 4]
```diff
@@ -227,21 +223,21 @@ export class MinaGuard extends SmartContract {
 
   // TODO: verify if we need an additional check here, to avoid front-running
   @method async setup(
```
[the remark this answers is not in the document]
Comment (graikos) [reply contracts/src/MinaGuard.ts]: Regarding list of owners, I don't see any gain from validating the commitment on chain. If we assume a malicious deployer/setup caller, then the client would need to compute the commitment off-chain and cross-check it themselves to make sure the correct owner list was used either way. Also, setting the commitment to arbitrary values (like all 0s) does not benefit a malicious deployer, as far as I can tell.

For the number of owners, I can't see any benefit either. At worst, it would brick the contract if it doesn't match the actual number. But this maybe needs more thought

[the remark this answers is not in the document]
Comment (mellowcroc) [reply contracts/src/MinaGuard.ts]: Agree that these don't necessarily need to be constraints. For list of owners, maybe we can just document that the commitment needs to be cross-checked by the consumer of the `setup` function.

**Review by mellowcroc — APPROVED** [review]
